### PR TITLE
[FIX] add move constructor and assignment operator

### PR DESCRIPTION
**Pull request by hroest**
# Description

adds move constructor / assignment operator to Targeted Experiment

- probably needs some testing

### CODE REVIEW

**Comment by timosachsenberg** [comment]
does =default not work here?

**Comment by timosachsenberg** [comment]
> does =default not work here?

ok I see we have some mutable members

**Comment by timosachsenberg** [comment]
> does =default not work here?

ok I see we have some mutable members. What about the reference maps? they don't get moved. what happens to them?

**Comment by timosachsenberg** [comment]
> does =default not work here?

ok I see we have some mutable members

**Comment by timosachsenberg** [comment]
> does =default not work here?

ok I see we have some mutable members

**Comment by hroest** [comment]
@timosachsenberg yes, they are not moved and that is why we mark these maps as dirty (need to be re-constructed before use). These are pointer-based maps, so they cannot be copied. We could argue that we should re-build them upon copy but that may be unnecessary work since we can do it when actually used

File: src/openms/source/ANALYSIS/TARGETED/TargetedExperiment.cpp [hunk 1]
```diff
@@ -98,6 +98,26 @@ namespace OpenMS
...
+    proteins_(std::move(rhs.proteins_)),
+    compounds_(std::move(rhs.compounds_)),
+    peptides_(std::move(rhs.peptides_)),
+    transitions_(std::move(rhs.transitions_)),
+    include_targets_(std::move(rhs.include_targets_)),
+    exclude_targets_(std::move(rhs.exclude_targets_)),
+    source_files_(std::move(rhs.source_files_)),
+    protein_reference_map_dirty_(true),
+    peptide_reference_map_dirty_(true),
+    compound_reference_map_dirty_(true)
```
Comment (hroest) [inline src/openms/source/ANALYSIS/TARGETED/TargetedExperiment.cpp:117]: @timosachsenberg the maps are empty and marked as "dirty" so they will not be used but rather re-built

**Comment by hroest** [comment]
@timosachsenberg merge?

**Comment by timosachsenberg** [comment]
> I think we could assume that the pointers stay valid when we move so we _could_ move the maps as well. What do you think? See https://stackoverflow.com/questions/25347599/am-i-guaranteed-that-pointers-to-stdvector-elements-are-valid-after-the-vector it seems that it is not strictly guaranteed since std::move _could_ be implemented through a copy

Yes, this is also what I saw in STL implementation - move is implemented like swap to keep iterators and pointers valid. I think if it had significant benefit (e.g, simplified the code, improve speed in typical situations or logic around marking the maps as dirty) we could do it. I

**Comment by cbielow** [comment]
You could remember the address of the first element before move and check if its identical after the move to avoid the dirty flag.

**Pull request merged**